docs/source/usage.rst

```diff
@@ -436,73 +434,85 @@ You can take advantage of nested interpolations to perform custom operations ove
-    ...                             args_as_strings=False)
+    >>> OmegaConf.new_register_resolver(
+    ...         "randint_nocache",
+    ...         lambda _, a, b: random.randint(a, b),
+    ...         use_cache=False)
     >>> c = OmegaConf.create({"x": "${randint_nocache:0, 1000}"})
     >>> c.x
     989
     >>> c.x  # not the same anymore since the cache is disabled
     796
 
-For more advanced operations on the config, you can allow a resolver to access the config
-object by setting `config_arg` -- preferably pointing to a keyword-only argument of
-the resolver.
-Note that you cannot use the cache in this case, since the resolver's output may now
-depend on arbitrary components of the config.
-
-.. doctest::
+For more advanced operations on the config, resolvers are given access to the parent
+of the node being processed through their first argument (which was not used in the
+examples above).
+Note that the cache ignores this parent argument (it only takes into account the values of the
+other resolver arguments). As a result it may be dangerous to use the cache *and* the parent,
+since the resolver’s output could now depend on arbitrary components of the config.
 
-    >>> OmegaConf.register_resolver("key_exists",
-    ...                             lambda key, *, config: key in config,
-    ...                             config_arg="config",
-    ...                             use_cache=False)
-    >>> c = OmegaConf.create({"a": 1,
-    ...                       "can_process_a": "${key_exists:a}",
-    ...                       "can_process_b": "${key_exists:b}"})
-    >>> c.can_process_a
-    True
-    >>> c.can_process_b
-    False
+The following example illustrates how this could be used to access a node's sibling
+(this is for illustration purpose only, a better way would be to use a relative interpolation):
 
-A similar mechanism can be used to access the parent of the current key being processed,
-by declaring a `parent_arg`. This may be useful to access keys in a relative way:
+.. doctest::
 
-    >>> OmegaConf.register_resolver(
-    ...         "get_sibling",
-    ...         lambda sibling, *, parent: getattr(parent, sibling),
-    ...         parent_arg="parent",
-    ...         use_cache=False)
+    >>> OmegaConf.new_register_resolver(
+    ...     "get_sibling",
+    ...     lambda parent, sibling: getattr(parent, sibling),
+    ...     use_cache=False)  # to always use the latest sibling value
     >>> c = OmegaConf.create(
     ...         {"foo": {"bar": {"baz1": 123,
     ...                          "baz2": "${get_sibling:baz1}"}}})
     >>> c.foo.bar.baz2
     123
+    >>> c.foo.bar.baz1 = 321
+    >>> c.foo.bar.baz2  # would be 123 if we had used `use_cache=True`
+    321
+
+If you would like to access the config's root and not just the parent node,
+use ``OmegaConf.get_root(parent)``.
+The following example shows how to obtain a boolean flag that indicates the
+presence of a specific key in the config:
```

Comment:
That's a nice example but a bit advanced.
I was going to suggest an example adding nodes, but it's already possible with nested interpolations as demonstrated above.

We can keep it, but if you think of a simpler example we can switch to it.

Reply:
Yeah, I struggled a bit with this. The simple examples I can think of actually don't need such an advanced mechanism and could be implemented in a simpler way...

Anyway, I gave it a shot in 526f5b7: this new example isn't really simpler, but at least I find it more interesting than the one I had before.

Reply:
The fact that we are having a hard time coming with a simple example showing this is useful may be an indication that it's not.
It's possible that by enabling nested interpolation support (along with the relative interpolation added in 2.1), we are already covered.

My vanilla example would be to show that you can make the value of one node a function of other nodes, but interpolation already allow for that.
```
node:
  x: 10
  y: 20
  sum: ${add:${.x}, ${.y}}
```
Not passing the parent here would potentially allow us to not have to deprecate register resolver, or at the very least make migration to the new function simpler.
There are other behavior changes that we need to think about though...

Reply:
> It's possible that by enabling nested interpolation support (along with the relative interpolation added in 2.1), we are already covered.

Yes I believe the combination of nested + relative interpolations covers most use cases I had in mind motivating gaining access to the config.

There's at least one important use case for it though: the env resolver needs it so that it can parse expressions containing interpolations (unless we decide not to support it).

I don't think we can avoid deprecating `register_resolver()` one way or another because of the change in the type of resolver arguments (e.g., `${resolver:1}` used to give `"1"` as input to the resolver, while I think it's clear that providing the integer `1` is better)

To be honest I'm not a fan of forcing the first argument of all resolvers to be the parent node, because it makes most resolvers look more complex than they need to be. Maybe I could change it to being an optional argument to `new_register_resolver()` instead? (this would bring us back closer to my original version, except that I wouldn't add another argument to access the root config since calling `OmegaConf.get_root(parent)` is easy enough, and we'd keep the new function name `new_register_resolver()` so as to avoid having too many arguments to `register_resolver()`, which was making it too bloated).

Reply:
> There's at least one important use case for it though: the env resolver needs it so that it can parse expressions containing interpolations (unless we decide not to support it).

Can you give an example of what you mean here?

> To be honest I'm not a fan of forcing the first argument of all resolvers to be the parent node, because it makes most resolvers look more complex than they need to be. Maybe I could change it to being an optional argument to new_register_resolver() instead

Do you mean to have a parameter to new_register_resolver that would make the callback receive the parent node if it's set to true?
```python
def new_register_resolver(..., pass_parent_node: bool = False):
 ...
```
I think that's not bad.

Reply:
Let's continue this thread of discussion in https://github.com/omry/omegaconf/pull/321#discussion_r527185026 since you proposed other options there. Still replying to your questions for the sake of completeness:

> Can you give an example of what you mean here?

I think you got it afterwards, but just in case, I was referring to something like the example in https://github.com/omry/omegaconf/pull/321#discussion_r472104910

> Do you mean to have a parameter to new_register_resolver that would make the callback receive the parent node if it's set to true?
> 
> ```python
> def new_register_resolver(..., pass_parent_node: bool = False):
>  ...
> ```

Pretty much, yes.